amethyst_rendy/src/bundle.rs

```diff
@@ -0,0 +1,950 @@
+        if self.outputs.is_none() {
+            return Err(format_err!(
+                "Trying to evaluate not fully defined pass {:?}. Missing `define_pass` call.",
+                self.key
+            ));
+        }
+        let mut outputs = self.outputs.unwrap();
+
+        ctx.mark_evaluating(self.key)?;
+
+        let mut target_ctx = TargetPlanContext {
+            plan_context: ctx,
+            key: self.key,
+            actions: vec![],
+            colors: outputs.colors.len(),
+            depth: outputs.depth.is_some(),
+            deps: vec![],
+        };
+
+        for extension in self.extensions {
+            extension(&mut target_ctx)?;
+        }
+
+        let TargetPlanContext {
+            mut actions, deps, ..
+        } = target_ctx;
+
+        let mut subpass = SubpassBuilder::new();
+        let mut pass = RenderPassNodeBuilder::new();
+
+        actions.sort_by_key(|a| a.0);
+        for action in actions.drain(..).map(|a| a.1) {
+            match action {
+                RenderableAction::RenderGroup(group) => {
+                    subpass.add_dyn_group(group);
+                }
+            }
+        }
+
+        for (i, color) in outputs.colors.drain(..).enumerate() {
+            match color {
+                OutputColor::Surface(surface, clear) => {
+                    subpass.add_color_surface();
+                    pass.add_surface(surface, clear);
+                }
+                OutputColor::Image(opts) => {
+                    let node = ctx.create_image(opts);
+                    ctx.register_output(TargetImage::Color(self.key, i), node)?;
+                    subpass.add_color(node);
+                }
+            }
+        }
+
+        if let Some(opts) = outputs.depth {
+            let node = ctx.create_image(opts);
+            ctx.register_output(TargetImage::Depth(self.key), node)?;
+            subpass.set_depth_stencil(node);
+        }
+
+        for node in deps {
+            subpass.add_dependency(node);
+        }
+
+        pass.add_subpass(subpass);
+        ctx.submit_pass(self.key, pass)?;
+        Ok(())
+    }
+}
+
+/// An action that represents a single transformation to the
+/// render graph, e.g. addition of single render group.
+///
+/// TODO: more actions needed for e.g. splitting pass into subpasses.
+#[derive(Debug)]
+pub enum RenderableAction<B: Backend> {
+    /// Register single render group for evaluation during target rendering
+    RenderGroup(Box<dyn RenderGroupBuilder<B, Resources>>),
+}
+
+impl<B: Backend> RenderableAction<B> {
+    fn colors(&self) -> usize {
+        match self {
+            RenderableAction::RenderGroup(g) => g.colors(),
+        }
+    }
+
+    fn depth(&self) -> bool {
+        match self {
+            RenderableAction::RenderGroup(g) => g.depth(),
+        }
+    }
+}
+
+/// Trait for easy conversion of various types into `RenderableAction` shell.
+pub trait IntoAction<B: Backend> {
```

Comment:
How about `impl<B, G> From<G> for RenderableActrion<B> where G: RenderGroupBuilder<B, Resources>>`?

Reply:
Orphan rule forbids that

Reply:
Ah, it collides with `From<T> for T`